compiler/lcalc/compile_without_exceptions.ml

```diff
@@ -125,420 +173,599 @@ let rec translate_typ (tau : typ) : typ =
+        })
+      scope_let_expr scope_let_next
+  | { scope_let_kind = SubScopeVarDefinition; scope_let_pos = pos; _ } ->
     Errors.raise_spanned_error pos
       "Internal Error: found an SubScopeVarDefinition that does not satisfy \
-       the invariants when translating Dcalc to Lcalc without exceptions: \
-       @[<hov 2>%a@]"
-      (Expr.format ctx.decl_ctx) expr
-  | ScopeLet
-      {
-        scope_let_kind = kind;
-        scope_let_typ = typ;
-        scope_let_expr = expr;
-        scope_let_next = next;
-        scope_let_pos = pos;
-      } ->
-    let var_is_pure =
-      match kind with
-      | DestructuringInputStruct -> (
-        (* Here, we have to distinguish between context and input variables. We
-           can do so by looking at the typ of the destructuring: if it's
-           thunked, then the variable is context. If it's not thunked, it's a
-           regular input. *)
-        match Marked.unmark typ with
-        | TArrow ([(TLit TUnit, _)], _) -> false
-        | _ -> true)
-      | ScopeVarDefinition | SubScopeVarDefinition | CallingSubScope
-      | DestructuringSubScopeResults | Assertion ->
-        true
-    in
-    let var, next = Bindlib.unbind next in
-    (* Cli.debug_print @@ Format.asprintf "unbinding %a" Print.var var; *)
-    let vmark = Expr.with_ty (Marked.get_mark expr) ~pos typ in
-    let ctx' = add_var vmark var var_is_pure ctx in
-    let new_var = (find ~info:"variable that was just created" var ctx').var in
-    Bindlib.box_apply2
-      (fun new_expr new_next ->
-        ScopeLet
-          {
-            scope_let_kind = kind;
-            scope_let_typ = translate_typ typ;
-            scope_let_expr = new_expr;
-            scope_let_next = new_next;
-            scope_let_pos = pos;
-          })
-      (Expr.Box.lift (translate_expr ctx ~append_esome:false expr))
-      (Bindlib.bind_var new_var (translate_scope_let ctx' next))
-
-let translate_scope_body
-    (scope_pos : Pos.t)
-    (ctx : 'm ctx)
-    (body : 'm D.expr scope_body) : 'm A.expr scope_body Bindlib.box =
-  match body with
+       the invariants when translating Dcalc to Lcalc without exceptions."
   | {
-   scope_body_expr = result;
-   scope_body_input_struct = input_struct;
-   scope_body_output_struct = output_struct;
+   scope_let_kind;
+   scope_let_typ;
+   scope_let_expr;
+   scope_let_next;
+   scope_let_pos;
   } ->
-    let v, lets = Bindlib.unbind result in
-    let vmark =
-      let m =
-        match lets with
-        | Result e | ScopeLet { scope_let_expr = e; _ } -> Marked.get_mark e
-      in
-      Expr.map_mark (fun _ -> scope_pos) (fun ty -> ty) m
+    (* base case *)
+    let next_var, next_body = Bindlib.unbind scope_let_next in
+
+    let next_var' = Var.translate next_var in
+    let ctx' =
+      Var.Map.add next_var
+        (match scope_let_kind with
+        | DestructuringInputStruct -> (
+          (* todo: we keep the old separation for further optimization while
```

Comment:
As you can notice, the branching here always give the same result ({info_pure=false; is_scope=false; var=next_var'}). but in the future, we could want to add different information, for instance if we notice that an expression is pure. Hence, it might be a good idea to keep this separation. 
I've added a comment about this.